<div style="background: #000;
            color: #FFF;
            margin: 0px;
                padding: 10px 0px 20px 0px;
            text-align: center; 
                ">
    <h1>Week 3 - Homework</h1>
</div>

<hr>

Exercise 1 due Monday.  

Exercise 2 due Tueday.

Exercise 3 due Wednesday.
<hr>

# Exercise 1

Write a for loop that sums the values of the following list and appends the different values to the correct list.

Remember what we learned about the value of an `int` and `str`.

Hint: try one part before the other.

In [75]:
# list to sum
list_vals = [82,True,'d',False,'v',7301,False,1,True,'x','F','w','Q','f',True]
#import hashlib;print("list_vals is correct" if hashlib.md5(str(list_vals).encode()).hexdigest() == '867f5d64395c44f016668cc442fb650c' else "list_vals is incorrect")


In [77]:
# creating empty lists
list_for_bools = []
list_for_ints = []
list_for_chars = []
#initilize this variable to hold the usm of the items in the list 
sum_of_all = 0
#looping through the original list 
for item in list_vals:
    # check if the item is boolean
    if type(item)==bool:
        list_for_bools.append(item)
        #keep tracking the sum
        sum_of_all +=int(item)
    # check if the item is int
    elif type(item)==int:
        list_for_ints.append(item)
         #keep tracking the sum
        sum_of_all +=item
    # check if the item is str and the len is 1 
    elif ((type(item)==str)&(len(item)==1)):
        list_for_chars.append(item)
         #keep tracking the sum
        sum_of_all +=ord(item)

In [78]:
# use the following to output your results:
print(list_for_bools)
print(list_for_ints)
print(list_for_chars)
print("The sum is ==",sum_of_all)
print("lenght of list_for_bools ==",len(list_for_bools))
print("lenght of list_for_ints == ", len(list_for_ints))
print("lenght of list_for_chars == ",len(list_for_chars))

[True, False, False, True, True]
[82, 7301, 1]
['d', 'v', 'x', 'F', 'w', 'Q', 'f']
The sum is == 8097
lenght of list_for_bools == 5
lenght of list_for_ints ==  3
lenght of list_for_chars ==  7


### Final Answer: 
`sum == 8097`  
`len(list_for_bools) == 5`  
`len(list_for_ints) == 3`   
`len(list_for_ints) == 7`   

# Exercise 2

Given a tuple of strings. Find the following value `x` and print it. The loop should end when `x` is found

In [79]:
x = '------'
# x='erw'
tuple_of_strings = ('steel', 'collect', 'capable', '------', 'shelter', 'language', 'mug', 'flowers')

# enter your code here:
check=False
for item in tuple_of_strings:
    if item == x:
        check=True
        break
if check:
    print("Element exist: ", x)
else:
    print("Element does not exist: ")

Element exist:  ------


# Exercise 3

Create a game in python. 

Rules:  
    1. You have 3 tries to guess a secret number.   
    2. Each try you are prompted to input a number.  
    3. You get a success message if you guess correctly and stop looping, else you subtract a try and go again.
    

In [31]:
##### use secret_number in your code but don't cheat
# importing needed libreries
import random
#generating a random integer number between 0 and 3
secret_number = random.randint(0,3)    

In [81]:
# numbers of times that user allow to repeat
tries=3
# keep tracking if we found the value
findIt=False

while tries>=1: # looping 
    user_try=int(input("Enter your guess number: ")) # promoting the user to enter guess number
    if user_try==secret_number: # getting good guess 
        print("Congrat! You got it") # printing out message 
        findIt=True # changing tracking to True
        break  # breaking out of loop
    else: # otherwise 
         print("Sorry, you did not get it. ") # user did not get it
    tries -=1 # keep decreaing the tries

# check if we did not find the value searched    
if not findIt:
    print("Sorry, you have finished all your tries. The secret number is:{} ".format(secret_number))



Enter your guess number: 3
Sorry, you did not get it. 
Enter your guess number: 2
Sorry, you did not get it. 
Enter your guess number: 0
Congrat! You got it


# Bonus Exercise
Research List & Dict Comprehension and try these problems. Also leave a written explanation of both.

a. 
Create a list comprehension that prints the following.
```python
[0, '*',2, '*', 4, '*', 6]

```

Hint: start from the most basic list comprehension  

In [53]:
#First method
a_list=[]  # creating an empty list
initial_list=list(range(0,7)) # creating a list of integers from 0 to 6
for element in initial_list: # looping through the list
    if element%2 !=0: # if the element is odd append to the list *
        a_list.append("*")
    else: # if the element is not odd, append element to the list
        a_list.append(element)
a_list #printing out the result       


[0, '*', 2, '*', 4, '*', 6]

In [54]:
#Second Method
text="0*2*4*6" # creating a text contains the desired output
a_list=[] # creating an empty list
for item in text: # looping through the list
    if item.isnumeric(): # checking if the element is numeric/digit
        a_list.append(int(item)) # if so, convert the element to integer
    else:
        a_list.append(item) # if not, keep the elemenet as it is
a_list #printing out the result 
        


[0, '*', 2, '*', 4, '*', 6]

b. Create a dict comprehension that prints the following:
```python
{'a': ['*', '*'], 'b': ['*', '*'],}

```


In [73]:
#first method
keys=['a','b'] # creating a list contains a and b
values=['*']*2 # creating a list contains two of * 
b_dictionnary={} # initializing a dictionnary
for key in keys: # looping through keys list 
    b_dictionnary[key]=values # assigning key to b_dictionary and values 
b_dictionnary # printing out the dictionary
    

{'a': ['*', '*'], 'b': ['*', '*']}

In [74]:
#second method: Direct way
b_dictionnary ={ # creating a dictionnary that has two keys a and b with an empty list 
    'a':[],
    'b':[]
}
for key in b_dictionnary: # looping through the keys of  dictionnary 
    # appending * to the values of each key in dictionary
    b_dictionnary[key].append('*') # append * to the list 
    b_dictionnary[key].append('*') # append * to the list 
b_dictionnary    
    

{'a': ['*', '*'], 'b': ['*', '*']}